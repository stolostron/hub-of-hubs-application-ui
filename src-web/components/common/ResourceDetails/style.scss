/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

@import '../../../../scss/variables';

#ResourceDetails {
  .app-info-and-dashboard-links {
    position: fixed;
    top: calc(5.1rem + var(--pf-c-page__main-section--PaddingTop));
    right: 12.75rem;
    z-index: 1001;
    background-color: $color__white;
    display: inline-block;
  }

  .app-info-and-dashboard-links-separator {
    border-left: 2px solid $color__gray-1;
    margin: 0 6.5rem 0 2rem;
  }

  .app-info-and-dashboard-links-separator-f {
    border-left: 2px solid $color__gray-1;
    margin: 0 0 0 0;
  }

  .app-info-icon {
    margin: 0 10px 0 0;
    vertical-align: sub;
  }

  .app-dashboard-icon {
    margin: 0 10px 0 0;
    vertical-align: sub;
  }

  .resource-diagram-title {
    font-size: 1.2rem;
    line-height: 1.25;
    font-weight: 300;
    padding: 20px 0 20px 0;
  }
}

.toastErrorMsg {
  margin-top: 2rem;
}

.toastSuccessMsg {
  margin-top: 2rem;
}
