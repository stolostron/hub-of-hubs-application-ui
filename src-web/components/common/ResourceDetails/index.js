/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
'use strict'

import React from 'react'
import { Route, Switch, withRouter, Redirect } from 'react-router-dom'
import { AcmAlert } from '@stolostron/ui-components'
import {
  updateSecondaryHeader,
  clearSuccessFinished
} from '../../../actions/common'
import { REQUEST_STATUS } from '../../../actions/index'
import {
  Alert,
  AlertGroup,
  AlertActionCloseButton
} from '@patternfly/react-core'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as Actions from '../../../actions'
import resources from '../../../../lib/shared/resources'
import msgs from '../../../../nls/platform.properties'
import ResourceOverview from '../ResourceOverview'

resources(() => {
  require('./style.scss')
})

const withResource = Component => {
  const mapDispatchToProps = dispatch => {
    return {
      actions: bindActionCreators(Actions, dispatch)
    }
  }

  const mapStateToProps = (state, ownProps) => {
    const { list: typeListName } = ownProps.resourceType,
          error = state[typeListName].err
    return {
      status: state[typeListName].status,
      statusCode: error && error.response && error.response.status
    }
  }

  return connect(mapStateToProps, mapDispatchToProps)(
    class extends React.PureComponent {
      static displayName = 'ResourceDetailsWithResouce';
      static propTypes = {
        actions: PropTypes.object,
        params: PropTypes.object,
        status: PropTypes.string,
        statusCode: PropTypes.object
      };

      constructor(props) {
        super(props)
        this.state = {
          errors: undefined
        }
      }

      componentDidMount() {
        // Clear the list of dropDowns
        const { params, actions } = this.props
        actions.clearAppDropDownList()
        // Then add it back so only one will be displaying
        actions.updateAppDropDownList(params.name)
      }
      render() {
        const { showError, errors } = this.state

        return (
          <React.Fragment>
            {showError && (
              <AcmAlert
                className="persistent"
                title={errors}
                variant="danger"
                noClose
                isInline
              />
            )}
            <Component {...this.props} />
          </React.Fragment>
        )
      }
    }
  )
}

const OverviewTab = withResource(ResourceOverview)

const components = {}

class ResourceDetails extends React.Component {
  constructor(props) {
    super(props)
    this.getBreadcrumb = this.getBreadcrumb.bind(this)
    this.handleErrorMsg = this.handleErrorMsg.bind(this)
    this.handleMutateClose = this.handleMutateClose.bind(this)

    this.otherBinding = {}
    const { routes } = this.props
    this.renderOverview = this.renderOverview.bind(this)
    routes.forEach(route => {
      this.otherBinding[route] = this.renderOther.bind(this, route)
    })

    this.state = {
      errorMsg: ''
    }
  }

  handleErrorMsg(err) {
    this.setState({ errorMsg: err })
  }

  handleMutateClose() {
    this.mutateFinished()
  }

  UNSAFE_componentWillMount() {
    const { updateSecondaryHeaderFn, tabs, launch_links, match } = this.props,
          params = match && match.params
    updateSecondaryHeaderFn(
      params.name,
      tabs,
      this.getBreadcrumb(),
      launch_links
    )
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.location !== this.props.location) {
      const { updateSecondaryHeaderFn, tabs, launch_links, match } = this.props,
            params = match && match.params

      updateSecondaryHeaderFn(
        params.name,
        tabs,
        this.getBreadcrumb(nextProps.location),
        launch_links
      )
    }
  }

  componentWillUnmount() {
    this.mutateFinished()
  }

  mutateFinished() {
    this.props.clearSuccessFinished()
  }

  render() {
    const { match, routes } = this.props
    return (
      <Switch>
        <Route exact path={match.url} render={this.renderOverview} />
        {routes &&
          routes.map(route => (
            <Route
              key={route}
              path={`${match.url}${route}`}
              render={this.otherBinding[route]}
            />
          ))}
        <Redirect to={match.url} />
      </Switch>
    )
  }

  renderOverview() {
    const {
      match,
      resourceType,
      staticResourceData,
      selectedNodeId,
      showExpandedTopology,
      actions,
      children,
      mutateStatus,
      mutateErrorMsg
    } = this.props
    const { locale } = this.context
    const { errorMsg } = this.state
    const syncSuccessKey = msgs.get(
      'dashboard.card.overview.cards.sync.success',
      locale
    )
    const syncErrorKey = msgs.get(
      'dashboard.card.overview.cards.sync.error',
      locale
    )
    return (
      <div id="ResourceDetails">
        {errorMsg && (
          <AlertGroup isToast className="toastErrorMsg">
            <Alert
              variant="danger"
              title={errorMsg}
              actionClose={
                <AlertActionCloseButton
                  title={errorMsg}
                  variantLabel="danger alert"
                  onClose={() => this.handleErrorMsg('')}
                />
              }
              key={errorMsg}
            />
          </AlertGroup>
        )}
        {mutateStatus === REQUEST_STATUS.ERROR && (
          <AlertGroup isToast className="toastErrorMsg">
            <Alert
              variant="danger"
              title={syncErrorKey}
              actionClose={
                <AlertActionCloseButton
                  title={syncErrorKey}
                  variantLabel="danger alert"
                  onClose={() => this.handleMutateClose()}
                />
              }
              key={syncErrorKey}
            >
              {mutateErrorMsg ||
                msgs.get(
                  'dashboard.card.overview.cards.sync.error.body',
                  locale
                )}
            </Alert>
          </AlertGroup>
        )}
        {mutateStatus === REQUEST_STATUS.DONE && (
          <AlertGroup isToast className="toastSuccessMsg">
            <Alert
              variant="success"
              title={syncSuccessKey}
              actionClose={
                <AlertActionCloseButton
                  title={syncSuccessKey}
                  variantLabel="success alert"
                  onClose={() => this.handleMutateClose()}
                />
              }
              key={syncSuccessKey}
            >
              {msgs.get(
                'dashboard.card.overview.cards.sync.success.body',
                locale
              )}
            </Alert>
          </AlertGroup>
        )}
        <OverviewTab
          resourceType={resourceType}
          params={match.params}
          staticResourceData={staticResourceData}
          actions={actions}
          modules={children}
          selectedNodeId={selectedNodeId}
          showExpandedTopology={showExpandedTopology}
          handleErrorMsg={this.handleErrorMsg}
        />
      </div>
    )
  }

  renderOther(route) {
    const {
      match,
      resourceType,
      staticResourceData,
      children,
      tabs
    } = this.props
    const Component = components[route]
    return (
      <Component
        resourceType={resourceType}
        params={match.params}
        tabs={tabs}
        staticResourceData={staticResourceData}
        modules={children}
      />
    )
  }

  getBreadcrumb(location) {
    location = location || this.props.location
    const { resourceType } = this.props,
          { locale } = this.context,
          urlSegments = location.pathname.replace(/\/$/, '').split('/')

    return [
      {
        label: msgs.get(`tabs.${resourceType.name.toLowerCase()}`, locale),
        url: urlSegments.slice(0, Math.min(3, urlSegments.length)).join('/')
      }
    ]
  }
}

ResourceDetails.contextTypes = {
  locale: PropTypes.string
}

ResourceDetails.propTypes = {
  actions: PropTypes.object,
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  clearSuccessFinished: PropTypes.func,
  launch_links: PropTypes.object,
  location: PropTypes.object,
  match: PropTypes.object,
  mutateErrorMsg: PropTypes.string,
  mutateStatus: PropTypes.string,
  resourceType: PropTypes.object,
  routes: PropTypes.array,
  selectedNodeId: PropTypes.string,
  showExpandedTopology: PropTypes.bool,
  staticResourceData: PropTypes.object,
  tabs: PropTypes.array,
  updateSecondaryHeaderFn: PropTypes.func
}

const mapDispatchToProps = dispatch => {
  return {
    actions: bindActionCreators(Actions, dispatch),
    updateSecondaryHeaderFn: (title, tabs, breadcrumbItems, links) =>
      dispatch(
        updateSecondaryHeader(title, tabs, breadcrumbItems, links, null, null)
      ),
    clearSuccessFinished: () => clearSuccessFinished(dispatch)
  }
}

const mapStateToProps = (state, ownProps) => {
  const { AppOverview } = state
  const { list: typeListName } = ownProps.resourceType,
        visibleResources = ownProps.getVisibleResources(state, {
          storeRoot: typeListName
        })

  const items = visibleResources.normalizedItems
  let params = {}
  if (ownProps.params) {
    params = ownProps.params
  } else {
    params = (ownProps.match && ownProps.match.params) || {}
  }
  const itemKey =
    (params &&
      params.name &&
      params.namespace &&
      decodeURIComponent(params.name) +
        '-' +
        decodeURIComponent(params.namespace)) ||
    undefined
  const item = (items && itemKey && items[itemKey]) || undefined
  const _uid = (item && item['_uid']) || ''
  const clusterName = (item && item['cluster']) || ''
  return {
    _uid,
    clusterName,
    selectedNodeId: AppOverview.selectedNodeId,
    showExpandedTopology: AppOverview.showExpandedTopology,
    params: params,
    mutateStatus: state[typeListName].mutateStatus,
    mutateErrorMsg: state[typeListName].mutateErrorMsg
  }
}

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ResourceDetails)
)
