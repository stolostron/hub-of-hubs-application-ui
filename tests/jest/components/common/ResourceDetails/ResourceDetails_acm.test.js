/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
jest.mock("../../../../../lib/client/apollo-client", () => ({
  getClient: jest.fn(() => {
    return null;
  }),
  getResource: jest.fn(resourceType => {
    const data = {
      data: {
        topology: {
          resources: [
            {
              id: "application--helloworld-local",
              type: "application",
              specs: {
                raw: {
                  apiVersion: "app.k8s.io/v1beta1",
                  kind: "Application",
                  metadata: {
                    name: "helloworld-local",
                    namespace: "acm"
                  },
                  spec: {}
                }
              }
            },
            {
              id: "application--helloworld-local",
              type: "service",
              specs: {
                raw: {
                  apiVersion: "v1",
                  kind: "Service",
                  metadata: {
                    name: "helloworld-app-svc",
                    namespace: "argo-helloworld"
                  }
                }
              }
            }
          ]
        }
      }
    };
    return Promise.resolve(data);
  }),
  get: jest.fn(resourceType => {
    const data = {
      data: {
        items: []
      }
    };
    return Promise.resolve(data);
  }),
  search: jest.fn(resourceType => {
    const data = {
      data: {
        searchResult: []
      }
    };
    return Promise.resolve(data);
  })
}));

jest.mock("../../../../../lib/client/access-helper.js", () => ({
  canCallAction: jest.fn(() => {
    const data = {
      data: {
        userAccess: {
          allowed: true
        }
      }
    };
    return Promise.resolve(data);
  }),
  canCreateActionAllNamespaces: jest.fn(() => {
    const data = {
      data: {
        userAccessAnyNamespaces: true
      }
    };
    return Promise.resolve(data);
  })
}));

const React = require("../../../../../node_modules/react");

import ResourceDetails from "../../../../../src-web/components/common/ResourceDetails";

import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import thunkMiddleware from "redux-thunk";

import {
  reduxStoreAppPipelineWithCEM_Inception,
  resourceType,
  staticResourceDataApp,
  HCMApplication,
  topology
} from "../../../components/TestingData";

const middleware = [thunkMiddleware];
const mockStore = configureMockStore(middleware);
const storeApp = mockStore(reduxStoreAppPipelineWithCEM_Inception);

const getVisibleResourcesFn = (state, store) => {
  const items = {
    normalizedItems: {
      "samplebook-gbapp-sample": HCMApplication
    }
  };
  return items;
};

const mockData = {
  getVisibleResources: getVisibleResourcesFn,
  location: {
    pathname: "/multicloud/applications/sample/samplebook-gbapp"
  },
  match: {
    isExact: true,
    path: "/multicloud/applications/:namespace/:name?",
    url: "/multicloud/applications/sample/samplebook-gbapp",
    params: {
      name: "samplebook-gbapp",
      namespace: "sample"
    }
  }
};

describe("ResourceDetails", () => {
  it("ResourceDetails renders correctly with data on single app with fectch resource on reload.", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={storeApp}>
            <ResourceDetails
              topology={topology}
              item={HCMApplication}
              match={mockData.match}
              loading={false}
              location={mockData.location}
              tabs={[]}
              routes={[]}
              params={mockData.match.params}
              getVisibleResources={mockData.getVisibleResources}
              resourceType={resourceType}
              staticResourceData={staticResourceDataApp}
            />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
