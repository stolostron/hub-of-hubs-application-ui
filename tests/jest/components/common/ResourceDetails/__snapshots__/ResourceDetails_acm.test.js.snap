// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceDetails ResourceDetails renders correctly with data on single app with fectch resource on reload. 1`] = `
<div
  id="ResourceDetails"
>
  
  <div
    className="overview-content"
    id="resource-overview"
  >
    <div
      className="app-info-and-dashboard-links"
    >
      <span
        className="app-info-and-dashboard-links-separator-f"
      />
    </div>
    <div
      className="overview-content-bottom overview-content-with-padding"
    >
      <div
        className="overview-cards-container"
      >
        <article
          className="pf-c-card pf-m-expanded"
          data-ouia-component-id="OUIA-Generated-Card-1"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe={true}
          id=""
        >
          <div
            className="pf-c-card__header makeStyles-header-2"
            onClick={[Function]}
          >
            <div
              className="pf-c-card__header-toggle"
            >
              <button
                aria-disabled={false}
                aria-expanded={true}
                aria-label="Toggle details"
                className="pf-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-1"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe={true}
                disabled={false}
                id="toggle-button"
                onClick={[Function]}
                role={null}
                type="button"
              >
                <span
                  className="pf-c-card__header-toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <div
              className="pf-c-card__title"
            >
              Details
            </div>
          </div>
          <div
            className="pf-c-card__expandable-content makeStyles-body-3"
          >
            <div
              className="pf-c-card__body"
            >
              <div
                className="pf-l-grid pf-m-all-12-col-on-sm pf-m-all-6-col-on-md"
              >
                <div
                  className="pf-l-grid__item makeStyles-leftCol-1"
                >
                  <dl
                    className="pf-c-description-list pf-m-horizontal"
                  >
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          Name
                           
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="app-name-container"
                          >
                            <div
                              className="app-name"
                            />
                            <div
                              className="pf-c-skeleton loading-skeleton-text"
                              style={
                                Object {
                                  "--pf-c-skeleton--Height": undefined,
                                  "--pf-c-skeleton--Width": "30%",
                                }
                              }
                            >
                              <span
                                className="pf-u-screen-reader"
                              />
                            </div>
                          </div>
                        </div>
                      </dd>
                    </div>
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          Namespace
                           
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </div>
                      </dd>
                    </div>
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          Created
                           
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </div>
                      </dd>
                    </div>
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                           
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "10%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </div>
                      </dd>
                    </div>
                  </dl>
                </div>
                <div
                  className="pf-l-grid__item"
                >
                  <dl
                    className="pf-c-description-list pf-m-horizontal"
                  >
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          Clusters
                           
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </div>
                      </dd>
                    </div>
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          Cluster resource status
                           
                          <svg
                            aria-hidden={true}
                            aria-labelledby={null}
                            className="help-icon"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            style={
                              Object {
                                "verticalAlign": "-0.125em",
                              }
                            }
                            viewBox="0 0 512 512"
                            width="1em"
                          >
                            <path
                              d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                            />
                          </svg>
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </div>
                      </dd>
                    </div>
                    <div
                      className="pf-c-description-list__group"
                    >
                      <dt
                        className="pf-c-description-list__term"
                      >
                        <span
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                           
                        </span>
                      </dt>
                      <dd
                        className="pf-c-description-list__description"
                      >
                        <div
                          className="pf-c-description-list__text"
                        >
                          <div
                            className="pf-c-skeleton loading-skeleton-text"
                            style={
                              Object {
                                "--pf-c-skeleton--Height": undefined,
                                "--pf-c-skeleton--Width": "30%",
                              }
                            }
                          >
                            <span
                              className="pf-u-screen-reader"
                            />
                          </div>
                        </div>
                      </dd>
                    </div>
                  </dl>
                </div>
              </div>
            </div>
          </div>
        </article>
        <div
          className="overview-cards-subs-section"
        >
          
          <button
            aria-disabled={true}
            aria-label={null}
            className="pf-c-button pf-m-secondary pf-m-disabled toggle-subs-btn"
            data-ouia-component-id="OUIA-Generated-Button-secondary-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            data-test-subscription-details={false}
            disabled={true}
            onClick={[Function]}
            role={null}
            tabIndex={null}
            type="button"
          >
            <div
              className="pf-c-skeleton loading-skeleton-text"
              style={
                Object {
                  "--pf-c-skeleton--Height": undefined,
                  "--pf-c-skeleton--Width": "70%",
                }
              }
            >
              <span
                className="pf-u-screen-reader"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      className="overview-content-bottom overview-content-with-padding"
    >
      <div
        className="resourceDiagramSourceContainer"
      >
        <div
          className="topology-controls"
        >
          <div
            className="topology-control-container"
          >
            <div
              id="search-textbox-portal-id"
            />
          </div>
        </div>
        <div
          className="resource-toolbar"
          id="resource-toolbar"
        >
          <div
            className="resource-toolbar-container"
          >
            <div
              className="resource-toolbar-buttons"
            >
              <div
                id="assorted-filter-open-portal-id"
              />
            </div>
            <div
              id="assorted-filter-close-portal-id"
            />
          </div>
        </div>
        <div
          className="resourceDiagramControlsContainer"
        >
          <div
            className="diagram-title"
          >
            Resource topology
            <svg
              className="diagram-title-divider"
            >
              <rect />
            </svg>
            <span
              className="how-to-read-text"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              How to read topology
              <svg
                className="how-to-read-icon"
              >
                <use
                  href="#diagramIcons_sidecar"
                />
              </svg>
            </span>
          </div>
          <div
            className="topologyDiagramContainer"
          >
            <div
              className="diagramViewerDiagram"
            >
              <div
                className="diagramViewerContainerContainer"
                id="diagram-viewer-container-container"
              >
                <div
                  aria-label="zoom"
                  className="diagramViewerContainer"
                  role="region"
                  style={
                    Object {
                      "height": "100%",
                      "width": "100%",
                    }
                  }
                >
                  <svg
                    className="topologyDiagram"
                    id="topologySvgId"
                    style={
                      Object {
                        "pointerEvents": "none",
                      }
                    }
                  />
                </div>
                <div
                  className="secondaryLoad"
                >
                  <span
                    aria-valuetext="Loading..."
                    className="pf-c-spinner pf-m-xl"
                    role="progressbar"
                  >
                    <span
                      className="pf-c-spinner__clipper"
                    />
                    <span
                      className="pf-c-spinner__lead-ball"
                    />
                    <span
                      className="pf-c-spinner__tail-ball"
                    />
                  </span>
                </div>
                <div
                  className="layoutLoadingContainer"
                >
                  <span
                    aria-valuetext="Loading..."
                    className="pf-c-spinner pf-m-xl"
                    role="progressbar"
                  >
                    <span
                      className="pf-c-spinner__clipper"
                    />
                    <span
                      className="pf-c-spinner__lead-ball"
                    />
                    <span
                      className="pf-c-spinner__tail-ball"
                    />
                  </span>
                </div>
              </div>
              <span
                className="diagramControls"
              >
                <div
                  className="diagram-controls"
                >
                  <div
                    aria-label="Zoom in"
                    className="zoom-in-button rectangle-zoom"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                    title="Zoom in"
                  >
                    <svg
                      className="icon"
                    >
                      <use
                        href="#diagramIcons_zoomIn"
                      />
                    </svg>
                  </div>
                  <div
                    aria-label="Zoom out"
                    className="zoom-out-button rectangle-zoom"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                    title="Zoom out"
                  >
                    <svg
                      className="icon"
                    >
                      <use
                        href="#diagramIcons_zoomOut"
                      />
                    </svg>
                  </div>
                  <div
                    aria-label="Zoom to fit"
                    className="zoom-target-button rectangle-zoom"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                    title="Zoom to fit"
                  >
                    <svg
                      className="icon_target"
                    >
                      <use
                        href="#diagramIcons_zoomTarget"
                      />
                    </svg>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
